### ISSUE TITLE
ddev 1.21.3/mutagen: file and directory permissions on host should match perms in container

### ISSUE THREAD

**Issue by garvinhicking** [drud/ddev#4323]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Expected Behavior

Files and directories synced over mutagen should carry over their permission bits

### Actual Behavior

Permissions are reset to only granting access to the owner, not for group/other

### Steps To Reproduce

Running latest macOS 12.6, a globally enabled mutagen project on startup syncs all files and directories with 0600 / 0700.

If I create a directory inside ddev, it's running usuall permissions according to umask (0644 / 0755). On the host this is synced back as "0600/0700". Creating files on Host with 0644 is synced to ddev with 0600.

### Anything else?

I am pretty sure, this worked in 1.12.1 before.

I run deployment scripts from within my ddev (using TYPO3 Surf). This uses rsync to sync a workspace directory. If I start the project where this workspace directory exists, all permissions are reset to 0600/0700, and deployed like this - on the server this leads to 403 errors because Apache does not deliver 0600/0700 files and dirs.

If I completely remove the workspace directory so that within ddev it is recreated from scratch, the permissions are alright. But of course the next time ddev is restarted, all permissions will be reset again. Also, they sync back to the host with wrong permissions.

Is this just a misconfiguration on my site? I definitely made deployments already with a ddev 1.21.1 and mutagen enabled, also with multiple restarts there. Was there some change I couldn't find?

Thanks!

**Comment by garvinhicking** [comment]
I'm sorry. I need to be productive for work, so I had to switch back to NFS ASAP; sadly I don't have a second machine to test this with.

I'll try to do that on the weekends.

I'm using TYPO3.Surf, not Deployer. Surf ist aimed for having zero dependencies on the host. So composer, frontend building, npm/grunt and other tasks are all performed locally, packaged, and uploaded to the host (through rsync). We have a lot of hosts, where we cannot perform building locally, so deployer is not really an option and would make building/packaging also not reliable, as it currently is (or not, due to the issue, lol).

Before I go down to the reproduction steps, a quick question: It should be possible, to have the same umask between mutagen client and mutagen server, yes? Are you aware of a change in 1.21.1 to 1.12.3 that could affect this permission issue?

Best regards,
Garvin

**Comment by rfay** [comment]
A new version of mutagen is in v1.21.3.

Mutagen does not attempt to make permissions exact, the reasoning is in https://github.com/mutagen-io/mutagen/issues/282

But if you demonstrate your exact problem in a reproducible way that does not involve rsync or deployment scripts, I'll be happy to work with you on it. I'm sure you know that with rsync (outgoing) you can set flags to change the destination permissions.

**Comment by garvinhicking** [comment]
Thank you. I know it's definitely up to me now to make this reproducible, I'll do my best to get this out soon.

I do know I could change rsync params (but thank you for pointing it out specifically), but we have many repositories using different configurations, and adding/patching different rsync into each repository would be a huge task. I'd rather find a solution with a mutagen option if possible.

**Comment by garvinhicking** [comment]
Addendum. Thanks for referencing the issue, I should've read this before posting. It describes my setup already quite well.

So I'll have to read up on --default-file-mode and --default-directory-mode and how to specifically run this. For our development, filemode 0644 and directorymode 0755 would be all we need. Any permissions that DIFFER from this on our deployment scenario is addressed through scripts running on the host. But we can't easily "fix" thousands of files and dirs transfered as 0600/0700.

I suppose I could add those default-file-mode/directory parameters to my global user mutagen configuration? No time left today, but I'll check this out tomorrow, if I find a solution, I'd happily see to create a PR for addings this in the mutagen ddev documentation to help others who may face the same situation.

**Comment by rfay** [comment]
Really, you'll want to either set rsync up with explicit perms or have a `chmod` and `chown` on your deployment anyway, because your deployment should never be dependent on the source-side ownership and permissions.

Regardless of that, I'm interested in seeing how this affects a *ddev* project in your case.

**Comment by garvinhicking** [comment]
Can you tell me, is there a global mutagen configuration that ddev uses (so: not per-project in .ddev/mutagen/mutagen.yml) but rather `~/.mutagen.yml`? The mutagen docs mention said file, but I wonder because ddev ships its own mutagen, is that still the right location? Or does it need to be a file like in `~/.ddev/mutagen.yml`?

If I set:

```
sync:
  defaults:
    permissions:
      defaultFileMode: 0644
      defaultDirectoryMode: 0700
```

I would be good.

(For now I'm just looking for a quick "get it running again", an adaptation of our deployment scripts to either use rsync params, or locally "fix" permissions is a bigger issue I'd rather not like to hot-tackle like my personal mutagen setup but needs more thorough team planning)

**Comment by rfay** [comment]
There is no global mutagen configuration. It would be nice if there were two levels of config for mutagen, but there aren't. 

DDEV sets the mutagen config to be .ddev/mutagen/mutagen.yml.

You can certainly make the change you suggest in .ddev/mutagen/mutagen.yml (and remove the #ddev-generated from the file so it doesn't get overwritten).

But you'll want `defaultDirectoryMode: 0755` I imagine.

Again... you don't really want source-side permissions to be how you control deployment-side-permissions. 

Also again, I'm interested in the actual problem, if described from a ddev-project point of view.

**Comment by garvinhicking** [comment]
Here's a longer writeup for my reasoning and reproducible situation. I'm being verbose to properly document this - not to waste anybody's time. I totally appreciate your time and effort, and this is my effort to give back to it, not to force a bookstack onto you. ;) 

By writing this, I think I found a workable way for me, so this journey was great at least for me.

# TL;DR 

I'll need to create a per-project `add_mutagen.sh` for me, to specify the needed mutagen options for `defaultFileMode`/`defaultDirectoryMode`. 

It would help me, if ddev would allow to specify a global mutagen file that would be concatenated with the ddev-generated one, if that ever is an option. I have zero knowledge in Go, but if you point me to a place in the ddev source, and how to build a customized ddev, I would try to provide a PR (more on this at the bottom of this comment).

# My ddev environment

As requested, a gist of my environment. I created a test repository here:

https://github.com/garvinhicking/ddev_deploy_mutagen

The `ddev debug test` output is there: https://github.com/garvinhicking/ddev_deploy_mutagen/blob/main/ddev-debug.txt

# The setup

Our goal is to change our project deployment from "per developer" to a central one. But we have some challenges to solve before we get there. So we have to deal with a non-ideal intermediate way.

For now, our setup is that each of our developers can check out a ddev project from git (we also have virtualbox+vagrant ones as legacy projects, which we usually convert to ddev whenever there's a deployment to perform). With very few steps, a repository can be setup so that each developer can perform the deployment:

```
ddev start
ddev exec "cd deployment && composer install && ./vendor/bin/surf deploy Production.Production"
```

# The deployment

What TYPO3 Surf (https://github.com/TYPO3/Surf) actually does is:

* It creates a "working space" (directory in `./deployment/workspace`), per target (i.e. one for Production, Staging, Development)
* It checks out a specified GIT branch into this directory tree (using a git working directory)
* It executes specified tasks: composer installation, frontend build (via npm/gulp), ...
* The resulting directory is then pushed via rsync to a target machine inside a temporary working directory, and can also exclude certain things (i.e. node_modules)
* On the server, Surf can execute defined tasks via SSH (i.e. `typo3cms` console for DB migrations, adding custom htaccess rules, downloading translation files, etc.). In the end, the working directory is then copied over to the final release directory structure.

The upside of this (compared to deployer) is:

* Near zero-dependencies on target host (only ssh/rsync)
* "Secure code is no code" - the target host will not contain possible build-related details, so if it gets hacked, "key build ingredients" can still be secure
* No git/credentials to the infrastructure are needed on the target host (i.e. private packagist keys, fontawesome serials, other code that only needs to exist in a compiled state on the target)
* Each developer can locally develop and "push" his project 

The downside of course is, that the target and development environments need to match up (same PHP versions).

In the future, a central build server would do the same a developer currently does locally, but on a controlled single server instance (which might even be running ddev for easier migration of our current projects).

# The problem

As you already mentioned, a problem is that since the build is created within the ddev environment (which is matched up with our target environment) and transferred via rsync, it pushes the current permissions of the ddev working space.

TYPO3 Surf uses `rsync -avz` for this for years:

https://github.com/TYPO3/Surf/blob/master/src/Task/RsyncFoldersTask.php#L73

So there's no way to "inject" custom rsync options like `--chmod` or `--perms` at this point. We'd have to create our own transport agent, which would mean to adapt possibly 100 repositories we currently use, all with different TYPO3 Surf and PHP versions and dependencies, so that would be a huge pain.

We could add a "post-rsync" hook and execute chown on files and directories after the fact, but again: We would need to adapt a hundred deployment configurations to add that hook.

So the bottom line is, it's not practical for us to fix this issue of permissions as it stands, but we need to satisfy the current condition, where the final local build needs to have the same 0644/0755 permissions as we need on the server side, plus maybe even 0600/0700 permissions for specific directories/files.

# Reproducing the problem

Now to describe what happens with mutagen, even though it's clear already with the issue you referenced.

* Spin up the mutagen project: `ddev start`
* Create test file: `ddev exec "cd deployment && touch file-created-in-ddev && ls -la file-created-in-ddev"`
* Result: `-rw-r--r-- 1 garvin dialout 0 Oct 25 09:10 file-created-in-ddev`
* Locally create a test file: `touch deployment/file-created-on-host && ls -la deployment/file-created-on-host`
* Result: `-rw-r--r--  1 garvin  staff  0 Oct 25 11:11 deployment/file-created-on-host`
* See the ddev-file on host: `ls -la deployment/file-created-in-ddev`
* Result: `-rw-------  1 garvin  staff  0 Oct 25 11:10 deployment/file-created-in-ddev`
* See the local test file on ddev: `ddev exec "ls -la deployment/file-created-on-host"`
* Result: `-rw------- 1 garvin dialout 0 Oct 25 09:11 deployment/file-created-on-host`

So as you can see, files created within ddev get the 0644/0755 perms. That would be great, and expected.

The files on the host though are transferred back as 0600/0700. I could live with that at this point, the host is of no concern to me.

However, inside the `workspace` directory, where Surf builds the files (this is not committed to GIT), it happens for me that after a `ddev poweroff` and maybe a reboot, that when I re-enter ddev, all the dirs/files in there have been changed to 0600/0700 permissions.

I couldn't reproduce this with the touched files above, maybe the files need to have content for mutagen to pick up synching and thus adjusting permissions.

# The solution

So, this problem with ddev and mutagen is a problem of "my own". Other developers use older ddev builds, or run Windows/WSL, and seem not be be affected by those permission problems.

That means, I need to find a solution to allow myself working with a ddev based project, without affecting others.

Changing the deployment pipeline would both be a lot of work for all projects I'm involved in, and also may affect the others.

First I wanted to fall back to NFS, but then I remembered that sadly on MaCOS a running nfsd means, powersaving of a macbook cannot be used.

My solution would cover:

* I want to use mutagen
* The project itself must NOT have mutagen enabled in its GIT configuration (not affect other developers)
* I want to add a non-default DDEV setting for the defaultFileMode / defaultDirectoryMode options (to not affect ddev users that are happy with current setup, which may be the more sensible default)
* Ideally I would want that for EVERY of our ddev projects, without needing to touch every project

Since there is no central `~/.mutagen.yaml` configuration that ddev uses (because the daemon runs per-ddev-project, as I understand), all of the above can currently not be reached (for now).

# Solution: Step 1

First, I need to make sure that mutagen is globally disabled, to not interfer (by creating a mutagen-config that is not of my own making).

Then I created a simple bash script `~/add_mutagen.sh` which enables mutagen per-project:

https://github.com/garvinhicking/ddev_deploy_mutagen/blob/main/add_mutagen.sh

I can call that script within each project once. The created files (`.ddev/config.mutagen.yml`, `.ddev/mutagen/mutagen.yml`) are in the `.gitignore` file, so they won't affect my co-developers. (Thinking about this, maybe only the `config.mutagen.yml` should be excluded?).

# Solution: Improving ddev?

To make this whole process easier, it would be awesome if ddev could allow to manage a central file that is merged with per-project mutagen creation. That would allow me to NOT maintain a customized mutagen configuration (per project), and benefit from future upstream patches to `mutagen.yml` which I would otherwise probably not even notice (because the file will no longer be ddev-maintained).

When ddev initializes its default `.ddev/mutagen/mutagen.yml` file, maybe it could look within `~/.ddev/global_mutagen.yml` and just prepend these settings to the autogenerated file? 

Then you could always overwrite global settings still with per-project settings again. Maybe even also add a possible `.ddev/global_mutagen_post.yml` file, which gets appended to the file, to ensure a way a global settings can always be global and never overwritten per-project. 

Both options would make sense to me, but I would understand if you do not want the maintenance overhead of this in the official ddev distro. If you're open to this, I'd be happy to try making a PR for ddev and try to understand the coding of ddev. Happy for any pointers of course, but I figure it shouldn't be too hard to make a lookup for two files and prepend/append them into an autogenerated file?

The End.

**Comment by rfay** [comment]
Are you running surf inside the web container or on the host?

I do find this problem with mutagen annoying, and think we should reopen the discussion with the maintainer of mutagen. 

However, can you not set permissions before you run your surf deployment?

**Comment by rfay** [comment]
Also, you're not willing to set permissions in .ddev/mutagen/mutagen.yml? Or that doesn't solve the problem?

**Comment by rfay** [comment]
I recommend that you enable mutagen globally (`ddev config global --mutagen-enabled`) rather than have a `config.mutagen.yaml`. BTW, you have `config.mutagen.yml`, if that works I'm surprised, as I *think* ddev just looks for `config.*.yaml`. 

Using global config for mutagen means that there will be nothing about it in your local config.

**Comment by garvinhicking** [comment]
I run surf inside the web container, not on the host. The host does not have the dependencies and PHP version as the web container, but that is required.

I can not set permissions before I run deployment, but that would mean changing the deployment recipe for 100 deployments, which would also affect my colleagues, which is a goalpoast I don't want to move.

In my lengthy eassy I did address that I cannot globally enable mutagen, because I need that directoryMode specifically set to 0644/0755, and I cannot do that globally from what I understand (you wrote "There is no global mutagen configuration."). So I'd need a per-project mutagen configuration.

If ddev would allow to have a global mutagen configuration that is applied, that would solve things for me of course.

(I tried it with config.mutagen.yml and not config.mutagen.yaml and it worked, yeah - I just verified that a second time)

**Comment by xenoscopic** [comment]
Regarding [this comment](https://github.com/mutagen-io/mutagen/issues/282#issuecomment-1290670685), I'm happy to re-open the discussion on additional permission propagation/preservation modes in Mutagen, but I'm not sure what the ask would be here.  It seems like the desire in this case is more about setting the default permission settings that exist, but if there's a specific behavior or configuration option that would help here, please let me know.

I'll open a new permissions propagation/preservation issue in any case, but is the specific proposal here just fully raw POSIX ownership/permission propagation?  UID/GIDs and perm/sticky/setuid/setgid bits?  Some subset of those?

**Comment by rfay** [comment]
Just raw POSIX permissions would be all I'd ask for. Ownership is too hard. sticky/setuid should never be used IMO. So just ugo/rwx

**Comment by xenoscopic** [comment]
Gotcha.  I'll go ahead and open up an initial proposal in the next day or two for what this should look like, trying to summarize the use cases and reference the prior discussion issues.

**Comment by fe-hicking** [comment]
Awesome, thank you. And just to add, yes - propagating ugo permissions is all I'd need, and would make my whole trip of adjusting the default permissions via mutagen configuration obsolete. Given that a possibly new permission option will be the ddev default ;-)

**Comment by garvinhicking** [comment]
@rfay Thanks. Yes, I actually use that for adjusting SSH key permissions.

I had to run a few tests before I got back here. Turns out my solution had a larger caveat that I recognized thanks to your message.

On a mutagen reset, of course all individual permissions within the container are reset. That includes things like `id_rsa` keys (which we copy from a NAS to our boxes for the deployment) or directories that TYPO3 sets to 0700 for internal usage. Also I was told a couple of tools and scripts we use actually set restricted permissions on files/dirs, and check those, and fail if permissions are not met.

Of course I could start to inventorize all possible special permission on special files and directories, and create a "fixup" script that I run time and again. But some of our projects contain a few tenthousand files, and doing that on every refresh feels pretty redundant.

I see that syncing the 'execute' permission bit carries over properly and "survive" a mutagen reset, so I assume some kind of check is in place. I see this touches the mutagen side of things and ddev can't do too much about it, but it would be oh-so-helpful if the raw ugo-Permissions could just be propagated from A<>B

**Comment by rfay** [comment]
Just a sidenote on your SSH keys. DDEV has `ddev auth ssh`, which is intended to be usable without ever copying keys into the container. You probably know that, but I wanted to make sure.

**Comment by garvinhicking** [comment]
(Thanks; actually we tried this when starting with ddev some time ago but faced some issues I can't remember; I think it was because we used company-global SSH keys that users should not operate with in their usual accounts, so our developers didnt have that key in our keychain, but only for the containers. Also we use a script that actually creates per-container unique deployment SSH keys through ssh-keygen, which we don't sync back to our host because it's not needed there)

**Comment by rfay** [comment]
I'm going to close this for now, but hopefully future improvements in mutagen will allow more options here. Thanks! I imagine you have everything sorted out now, if not, happy to continue the conversation.

**Comment by garvinhicking** [comment]
@rfay Thank you. Yeah I got that cumbersome script working.

I also opened an issue with the mutagen project.

However, I still remain curious if ddev couldn't provide the ability to globally specify mutagen options. Would you be open for a more isolated/brief issue request to track that feature?

If I could just have a central yaml file that gets merged with each mutagen configuration of each ddev docker instance, it would make my hackish way of applying it per-project obselte.

Best regards,
Garvin

**Comment by rfay** [comment]
@garvinhicking I have talked with @xenoscopic about an option for a ddev-global mutagen config, and it turns out that

1. It would already be possible for DDEV to use the ~/.mutagen.yml global file... except that DDEV explicitly creates sync sessions with `--no-global-configuration`. I don't think I'll change that, because we don't want global mutagen config intended for other things to accidentally interfere with ddev usage.
2. @xenoscopic and I did talk about another option for config, and he implied he might be able to slip in multiple config files (allowing a DDEV global mutagen.yml) in a coming release. It might help for you to open an issue in the mutagen queue for that, even though it's on his (informal) radar.
